plugins/lookup/azure_keyvault_secret.py

```diff
@@ -0,0 +1,178 @@
+        secret:
+            description: Secret of the service principal.
+        tenant_id:
+            description: Tenant id of service principal.
+    notes:
+        - If version is not provided, this plugin will return the latest version of the secret.
+        - If ansible is running on Azure Virtual Machine with MSI enabled, client_id, secret and tenant isn't required.
+        - For enabling MSI on Azure VM, please refer to this doc https://docs.microsoft.com/en-us/azure/active-directory/managed-service-identity/
+        - After enabling MSI on Azure VM, remember to grant access of the Key Vault to the VM by adding a new Acess Policy in Azure Portal.
+        - If MSI is not enabled on ansible host, it's required to provide a valid service principal which has access to the key vault.
+        - To use a plugin from a collection, please reference the full namespace, collection name, and lookup plugin name that you want to use.
+"""
+
+EXAMPLE = """
+- name: Look up secret when ansible host is MSI enabled Azure VM
+  debug: msg="the value of this secret is {{lookup('azure.azcollection.azure_keyvault_secret','testSecret/version',vault_url='https://yourvault.vault.azure.net')}}"
+
+- name: Look up secret when ansible host is general VM
+  vars:
+    url: 'https://yourvault.vault.azure.net'
+    secretname: 'testSecret/version'
+    client_id: '123456789'
+    secret: 'abcdefg'
+    tenant: 'uvwxyz'
+  debug: msg="the value of this secret is {{lookup('azure.azcollection.azure_keyvault_secret',secretname,vault_url=url, cliend_id=client_id, secret=secret, tenant_id=tenant)}}"
+
+# Example below creates an Azure Virtual Machine with SSH public key from key vault using 'azure_keyvault_secret' lookup plugin.
+- name: Create Azure VM
+  hosts: localhost
+  connection: local
+  no_log: True
+  vars:
+    resource_group: myResourceGroup
+    vm_name: testvm
+    location: eastus
+    ssh_key: "{{ lookup('azure.azcollection.azure_keyvault_secret','myssh_key') }}"
+  - name: Create VM
+    azure_rm_virtualmachine:
+      resource_group: "{{ resource_group }}"
+      name: "{{ vm_name }}"
+      vm_size: Standard_DS1_v2
+      admin_username: azureuser
+      ssh_password_enabled: false
+      ssh_public_keys:
+        - path: /home/azureuser/.ssh/authorized_keys
+          key_data: "{{ ssh_key }}"
+      network_interfaces: "{{ vm_name }}"
+      image:
+        offer: UbuntuServer
+        publisher: Canonical
+        sku: 16.04-LTS
+        version: latest
+"""
+
+RETURN = """
+  _raw:
+    description: secret content string
+"""
+
+from ansible.errors import AnsibleError, AnsibleParserError
+from ansible.plugins.lookup import LookupBase
+from ansible.utils.display import Display
+import requests
+
+display = Display()
+
+TOKEN_ACQUIRED = False
+
+token_params = {
+    'api-version': '2018-02-01',
+    'resource': 'https://vault.azure.net'
+}
+token_headers = {
+    'Metadata': 'true'
+}
+token = None
+try:
```

Comment:
When plugins doesn't retrieve MSI Token, we must wait 30 seconds by lookup. I've forked this plugins to comment this code because my Ansible Agent can't use MSI Token.

Reply:
Timeout changed as suggested by @ttsakpc